NullPointerException when using spring-boot-devtools

<!--- This template is for bugs. Remove it for other issues. -->

<!--- Choose an expressive title -->
### Description
Created the demo project using the spring-boot maven archetype:
`mvn archetype:generate "-DarchetypeGroupId=com.blazebit" "-DarchetypeArtifactId=blaze-persistence-archetype-spring-boot-sample" "-DarchetypeVersion=1.5.1"`

Add the spring-boot-devtools dependency:
```
<dependency>
     <groupId>org.springframework.boot</groupId>
     <artifactId>spring-boot-devtools</artifactId>
     <optional>true</optional>
</dependency>
```

### Expected behavior
The Application should startup.

### Actual behavior
Application startup fails with NullPointerException
```
Caused by: java.lang.NullPointerException: null
	at com.blazebit.persistence.view.impl.metamodel.ViewMetamodelImpl.<init>(ViewMetamodelImpl.java:117) ~[blaze-persistence-entity-view-impl-1.5.1.jar:1.5.1]
	at com.blazebit.persistence.view.impl.EntityViewManagerImpl.<init>(EntityViewManagerImpl.java:261) ~[blaze-persistence-entity-view-impl-1.5.1.jar:1.5.1]
```

### Steps to reproduce
<!--- Give us enough details so we can create a testcase that reproduces this problem -->
<!--- Idealy you would create a test case based on the template project from: https://github.com/Blazebit/blaze-persistence-test-case-template -->
<!--- Either attach a ZIP containing the test case or create a pull request against the test-case-template repository and put the link to the PR here -->

### Environment
<!--- Environment info like e.g.					-->
<!--- Version:            1.2.0-Alpha1				-->
<!--- JPA-Provider:       Hibernate 5.2.7.Final		-->
<!--- DBMS:               PostgreSQL 9.6.2			-->
<!--- Application Server: Java SE					-->

Version:         1.5.1    
JPA-Provider:       
DBMS:               
Application Server: Spring Boot

This is due to the fact that the Spring Devtools RestartClassLoader instantiates another Class instance for the entities.

When `EntityModelImpl.getManagedTtype` the `Class` instance which is passed as an argument, for the same domain-type, will be a different `Class` instance. This will cause `.classMap.get()` to return `null`.

The bug is inside `EntityViewConfigurationDelegate`. 
Instead of using:
`Class<?> clazz = Class.forName(candidate.getBeanClassName());`

You should use the `ClassUtils` from Spring itself which will gracefully recognize Spring Boot's `RestartClassLoader`.

Something like this seems to work for me locally:
`Class<?> clazz = ClassUtils.forName( candidate.getBeanClassName(),  resourceLoader == null ? null : resourceLoader.getClassLoader() );`

I will see if I can create a PR for this, however a test case is a bit complicated due to the fact it includes classloaders.

--- a/integration/entity-view-spring/src/main/java/com/blazebit/persistence/integration/view/spring/impl/EntityViewConfigurationDelegate.java
+++ b/integration/entity-view-spring/src/main/java/com/blazebit/persistence/integration/view/spring/impl/EntityViewConfigurationDelegate.java
@@ -22,6 +22,7 @@ import org.springframework.beans.factory.support.BeanDefinitionRegistry;
 import org.springframework.beans.factory.support.GenericBeanDefinition;
 import org.springframework.core.env.Environment;
 import org.springframework.core.io.ResourceLoader;
+import org.springframework.util.ClassUtils;
 
 import java.util.HashSet;
 import java.util.Set;
@@ -48,7 +49,7 @@ public class EntityViewConfigurationDelegate {
         Set<Class<?>> entityViewListenerClasses = new HashSet<>();
         for (BeanDefinition candidate : configurationSource.getCandidates(resourceLoader)) {
             try {
-                Class<?> clazz = Class.forName(candidate.getBeanClassName());
+                Class<?> clazz = ClassUtils.forName( candidate.getBeanClassName(),  resourceLoader == null ? null : resourceLoader.getClassLoader() );
                 if (clazz.isAnnotationPresent(EntityView.class)) {
                     entityViewClasses.add(clazz);
                 } else {

--- a/integration/entity-view-spring/src/test/java/com/blazebit/persistence/integration/view/spring/AnnotationClassLoaderTest.java
+++ b/integration/entity-view-spring/src/test/java/com/blazebit/persistence/integration/view/spring/AnnotationClassLoaderTest.java
@@ -1,0 +1,89 @@
+/*
+ * Copyright 2014 - 2021 Blazebit.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.blazebit.persistence.integration.view.spring;
+
+import com.blazebit.persistence.integration.view.spring.views.sub1.TestView1;
+import com.blazebit.persistence.integration.view.spring.views.sub2.TestView2;
+import com.blazebit.persistence.view.spi.EntityViewConfiguration;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.springframework.cglib.proxy.Enhancer;
+import org.springframework.cglib.proxy.MethodInterceptor;
+import org.springframework.cglib.proxy.MethodProxy;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+
+import javax.inject.Inject;
+import java.lang.reflect.Method;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * @author Moritz Becker
+ * @since 1.2.0
+ */
+@RunWith(SpringJUnit4ClassRunner.class)
+@ContextConfiguration(classes = AnnotationClassLoaderTest.TestConfig.class
+)
+public class AnnotationClassLoaderTest {
+
+    static final AtomicInteger trackedClassLoaderViews = new AtomicInteger();
+    static final ClassLoader previousClassLoader = Thread.currentThread().getContextClassLoader();
+
+    static {
+        Enhancer enhancer = new Enhancer();
+        enhancer.setSuperclass(ClassLoader.class);
+        enhancer.setCallback(new MethodInterceptor() {
+            @Override
+            public Object intercept(Object object, Method method, Object[] args, MethodProxy proxy) throws Throwable {
+                if (method.getName().contains("loadClass") && args.length == 1) {
+                    if (Objects.equals(TestView1.class.getName(), args[0]) || Objects.equals(TestView2.class.getName(), args[0])) {
+                        trackedClassLoaderViews.incrementAndGet();
+                    }
+                }
+                return proxy.invokeSuper(object, args);
+            }
+        });
+        ClassLoader proxy = (ClassLoader) enhancer.create();
+        Thread.currentThread().setContextClassLoader(proxy);
+    }
+
+    @AfterClass
+    public static void verify() {
+        Thread.currentThread().setContextClassLoader(previousClassLoader);
+    }
+
+    @Inject
+    private EntityViewConfiguration entityViewConfiguration;
+
+    @Test
+    public void testInjectionWithCustomLoader() {
+        Set<Class<?>> entityViews = entityViewConfiguration.getEntityViews();
+        Assert.assertEquals(2, entityViews.size());
+        Assert.assertEquals(2, trackedClassLoaderViews.get());
+        Assert.assertTrue(entityViewConfiguration.getEntityViews().contains(TestView1.class));
+        Assert.assertTrue(entityViewConfiguration.getEntityViews().contains(TestView2.class));
+    }
+
+    @Configuration
+    @EnableEntityViews
+    static class TestConfig {
+    }
+}

--- a/integration/entity-view-spring/src/test/java/com/blazebit/persistence/integration/view/spring/AnnotationClassLoaderTest.java
+++ b/integration/entity-view-spring/src/test/java/com/blazebit/persistence/integration/view/spring/AnnotationClassLoaderTest.java
@@ -22,15 +22,11 @@ import org.junit.AfterClass;
 import org.junit.Assert;
 import org.junit.Test;
 import org.junit.runner.RunWith;
-import org.springframework.cglib.proxy.Enhancer;
-import org.springframework.cglib.proxy.MethodInterceptor;
-import org.springframework.cglib.proxy.MethodProxy;
 import org.springframework.context.annotation.Configuration;
 import org.springframework.test.context.ContextConfiguration;
 import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
 
 import javax.inject.Inject;
-import java.lang.reflect.Method;
 import java.util.Objects;
 import java.util.Set;
 import java.util.concurrent.atomic.AtomicInteger;
@@ -48,25 +44,11 @@ public class AnnotationClassLoaderTest {
     static final ClassLoader previousClassLoader = Thread.currentThread().getContextClassLoader();
 
     static {
-        Enhancer enhancer = new Enhancer();
-        enhancer.setSuperclass(ClassLoader.class);
-        enhancer.setCallback(new MethodInterceptor() {
-            @Override
-            public Object intercept(Object object, Method method, Object[] args, MethodProxy proxy) throws Throwable {
-                if (method.getName().contains("loadClass") && args.length == 1) {
-                    if (Objects.equals(TestView1.class.getName(), args[0]) || Objects.equals(TestView2.class.getName(), args[0])) {
-                        trackedClassLoaderViews.incrementAndGet();
-                    }
-                }
-                return proxy.invokeSuper(object, args);
-            }
-        });
-        ClassLoader proxy = (ClassLoader) enhancer.create();
-        Thread.currentThread().setContextClassLoader(proxy);
+        Thread.currentThread().setContextClassLoader(new TrackingClassLoader());
     }
 
     @AfterClass
-    public static void verify() {
+    public static void resetClassLoader() {
         Thread.currentThread().setContextClassLoader(previousClassLoader);
     }
 
@@ -85,5 +67,15 @@ public class AnnotationClassLoaderTest {
     @Configuration
     @EnableEntityViews
     static class TestConfig {
+    }
+
+    private static class TrackingClassLoader extends ClassLoader {
+        @Override
+        public Class<?> loadClass(String name) throws ClassNotFoundException {
+            if (Objects.equals(TestView1.class.getName(), name) || Objects.equals(TestView2.class.getName(), name)) {
+                trackedClassLoaderViews.incrementAndGet();
+            }
+            return super.loadClass(name);
+        }
     }
 }

